packages/flutter/lib/src/material/time_picker.dart

```diff
@@ -136,77 +485,147 @@ enum _TimePickerHeaderId {
+    double tailWidth = childSizes[fragments[pivotIndex].layoutId].width / 2.0;
+    for (_TimePickerHeaderFragment fragment in fragments.skip(pivotIndex + 1)) {
+      tailWidth += childSizes[fragment.layoutId].width + fragment.leftMargin;
+    }
 
-        x -= hourSize.width;
-        positionChild(_TimePickerHeaderId.hour, new Offset(x, timeCenter - hourSize.height / 2.0));
+    double x = width / 2.0 + tailWidth;
+    x = math.min(x, width);
+    for (int i = fragments.length - 1; i >= 0; i -= 1) {
+      final _TimePickerHeaderFragment fragment = fragments[i];
+      final Size childSize = childSizes[fragment.layoutId];
+      x -= childSize.width;
+      positionChild(fragment.layoutId, new Offset(x, y - childSize.height / 2.0));
+      x -= fragment.leftMargin;
+    }
+  }
 
-        x = (size.width - periodSize.width) / 2.0;
-        positionChild(_TimePickerHeaderId.period, new Offset(x, timeCenter + timeHeight / 2.0 + _kPeriodGap));
-        break;
+  void _positionPiece(double width, double centeredAroundY, Map<_TimePickerHeaderId, Size> childSizes, List<_TimePickerHeaderFragment> fragments) {
+    double pieceWidth = 0.0;
+    double nextMargin = 0.0;
+    for (_TimePickerHeaderFragment fragment in fragments) {
+      final Size childSize = childSizes[fragment.layoutId];
+      pieceWidth += childSize.width + nextMargin;
+      // Delay application of margin until next element because margin of the
+      // left-most fragment should not contribute to the size.
+      nextMargin = fragment.leftMargin;
+    }
+    double x = (width + pieceWidth) / 2.0;
+    for (int i = fragments.length - 1; i >= 0; i -= 1) {
+      final _TimePickerHeaderFragment fragment = fragments[i];
+      final Size childSize = childSizes[fragment.layoutId];
+      x -= childSize.width;
+      positionChild(fragment.layoutId, new Offset(x, centeredAroundY - childSize.height / 2.0));
+      x -= fragment.leftMargin;
     }
   }
 
   @override
-  bool shouldRelayout(_TimePickerHeaderLayout oldDelegate) => orientation != oldDelegate.orientation;
+  bool shouldRelayout(_TimePickerHeaderLayout oldDelegate) => orientation != oldDelegate.orientation || format != oldDelegate.format;
 }
 
+/// Provides properties for rendering time picker header fragments.
+@immutable
+class _TimePickerFragmentContext {
+  const _TimePickerFragmentContext({
+    @required this.headerTextTheme,
+    @required this.selectedTime,
+    @required this.mode,
+    @required this.activeColor,
+    @required this.activeStyle,
+    @required this.inactiveColor,
+    @required this.inactiveStyle,
+    @required this.onTimeChange,
+    @required this.onModeChange,
+  });
+
+  final TextTheme headerTextTheme;
+  final TimeOfDay selectedTime;
+  final _TimePickerMode mode;
+  final Color activeColor;
+  final TextStyle activeStyle;
+  final Color inactiveColor;
+  final TextStyle inactiveStyle;
+  final ValueChanged<TimeOfDay> onTimeChange;
+  final ValueChanged<_TimePickerMode> onModeChange;
+}
```

Comment:
put this class before the fragments, so that reading the code you see this first